enzyme/Enzyme/InstructionBatcher.h

```diff
@@ -0,0 +1,242 @@
+//===- InstructionBatcher.h
+//--------------------------------------------------===//
+//
+//                             Enzyme Project
+//
+// Part of the Enzyme Project, under the Apache License v2.0 with LLVM
+// Exceptions. See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// If using this code in an academic setting, please cite the following:
+// @incollection{enzymeNeurips,
+// title = {Instead of Rewriting Foreign Code for Machine Learning,
+//          Automatically Synthesize Fast Gradients},
+// author = {Moses, William S. and Churavy, Valentin},
+// booktitle = {Advances in Neural Information Processing Systems 33},
+// year = {2020},
+// note = {To appear in},
+// }
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains an instruction visitor InstructionBatcher that generates
+// the batches all LLVM instructions.
+//
+//===----------------------------------------------------------------------===//
+
+#include "llvm/IR/InstVisitor.h"
+
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/SmallVector.h"
+
+#include "llvm/Support/Casting.h"
+
+#include "llvm/IR/Constants.h"
+#include "llvm/IR/IRBuilder.h"
+#include "llvm/IR/Value.h"
+
+#include "llvm/Transforms/Utils/BasicBlockUtils.h"
+#include "llvm/Transforms/Utils/Cloning.h"
+#include "llvm/Transforms/Utils/ValueMapper.h"
+
+#include "GradientUtils.h"
+
+using namespace llvm;
+
+class InstructionBatcher : public llvm::InstVisitor<InstructionBatcher> {
+public:
+  InstructionBatcher(
+      Function *oldFunc, Function *newFunc, unsigned width,
+      ValueMap<const Value *, std::vector<Value *>> &vectorizedValues,
+      ValueToValueMapTy &originalToNewFn, SmallPtrSetImpl<Value *> &toVectorize,
+      EnzymeLogic &Logic)
+      : vectorizedValues(vectorizedValues), originalToNewFn(originalToNewFn),
+        toVectorize(toVectorize), width(width), Logic(Logic) {}
+
+private:
+  ValueMap<const Value *, std::vector<Value *>> &vectorizedValues;
+  ValueToValueMapTy &originalToNewFn;
+  SmallPtrSetImpl<Value *> &toVectorize;
+  unsigned width;
+  EnzymeLogic &Logic;
+
+public:
+  void visitInstruction(llvm::Instruction &inst) {
+    auto found = vectorizedValues.find(&inst);
+    assert(found != vectorizedValues.end());
+    auto placeholders = found->second;
+    Instruction *placeholder = cast<Instruction>(placeholders[0]);
+
+    for (unsigned i = 1; i < width; ++i) {
+      ValueToValueMapTy vmap;
+      Instruction *new_inst = placeholder->clone();
+      vmap[placeholder] = new_inst;
+
+      for (unsigned j = 0; j < inst.getNumOperands(); ++j) {
+        Value *op = inst.getOperand(j);
+        Value *new_op;
+        if (isa<Constant>(op)) {
```

Comment:
Can we extract this into a function to evaluate a condition? If it's a constant data I concur, but if this is a global, you might need a batch global or something?